File: MultilanguageSDK/Python/Samples/OmniGuard/SafetyCameraGetResult.py
from pickle import FALSE, TRUE
import sys
currentPath =  sys.path[0]
pos = currentPath.find('Samples')
libpath = currentPath[:pos]
print(libpath + "API")
sys.path.append(libpath) #absolutely path

from API.ScepterDS_api import *
import time
from typing import List
import struct

class AlgHeader(Structure):
    _pack_ = 1
    _fields_ = [
        ("startIdentifer", c_uint32),
        ("sizeOfValidData", c_uint16),
        ("checksum", c_uint8)
    ]
class DetectedResult(Structure):
    _pack_ = 1
    _fields_ = [
        ("index", c_uint8),
        ("safetyLevel", c_uint8),
        ("isWarning", c_uint8)
    ]
print("---SafetyCameraGetResult---")
camera = ScepterTofCam()
def parse_ai_result(result: ScAIResult):
    detected_results = []
    used_len = 0
    
    try:
        # 1. check data length
        header_size = ctypes.sizeof(AlgHeader)
        if result.dataLen < header_size:
            print(f"[ERROR] Insufficient data length: {result.dataLen} bytes (header requires {header_size} bytes)")
            return -1, []
        
        # 2. parse header
        header = AlgHeader.from_buffer_copy(
            ctypes.string_at(result.pResultData, header_size))
        
        if header.startIdentifer != 0x474C4102:
            print(f"[ERROR] Invalid header identifier: 0x{header.startIdentifer:08X} (expected 0x474C4102)")
            return -1, []
        
        used_len += header_size
        
        # 3. check vaild data length
        if result.dataLen < used_len + header.sizeOfValidData:
            print(f"[ERROR] Invalid data length: {header.sizeOfValidData} bytes specified but only {result.dataLen - used_len} bytes available")
            return -1, []
        
        # 4. vaild checksum
        data_slice = ctypes.string_at(
            ctypes.byref(result.pResultData.contents, used_len),
            header.sizeOfValidData)
        calculated_checksum = sum(data_slice) & 0xFF
        
        if calculated_checksum != header.checksum:
            print(f"[ERROR] Checksum mismatch: calculated 0x{calculated_checksum:02X}, expected 0x{header.checksum:02X}")
            return -1, []
        
        # 5. parse PackageID
        package_id = data_slice[0]
        used_len += 1
        if package_id != 0x00:
            print(f"[ERROR] Invalid package ID: 0x{package_id:02X} (expected 0x00)")
            return -1, []
        
        # 6. parse result length
        result_len = struct.unpack_from('<H', data_slice, 1)[0]
        used_len += 2
        
        # 7. parse state and error code
        state = data_slice[3]
        error_code = data_slice[4]
        used_len += 2
        
        if state != 0 or error_code != 0:
            print(f"[ERROR] State check failed - state: 0x{state:02X}, errorCode: 0x{error_code:02X}")
            return -1, []
        
        # 8. parse area count
        area_count = data_slice[5]
        used_len += 1
        
        # 9. parse detect result
        result_size = ctypes.sizeof(DetectedResult)
        offset = 6  # offset
        
        for _ in range(area_count):
            if offset + result_size > len(data_slice):
                print(f"[WARNING] Truncated data at result {len(detected_results) + 1}/{area_count}")
                break
            
            dr = DetectedResult.from_buffer_copy(data_slice, offset)
            detected_results.append(dr)
            offset += result_size
            used_len += result_size
        
        # 10. check the rest data
        if result.dataLen > used_len:
            remaining = result.dataLen - used_len
            print(f"[INFO] Result {result.resultIndex} has {remaining} bytes remaining data.")
        
        return 0, detected_results
    
    except Exception as e:
        print(f"[CRITICAL] Parsing failed: {str(e)}")
        return -1, []


camera_count = camera.scGetDeviceCount(3000)
print("[scGetDeviceCount] success, ScStatus(0), The device count is {}".format(camera_count))
if camera_count <= 0:
    print("[scGetDeviceCount] scans for 3000ms and then returns the device count is 0. Make sure the device is on the network before running the samples.")
    exit()

device_info=ScDeviceInfo()

if camera_count > 0:
    ret,device_infolist=camera.scGetDeviceInfoList(camera_count)
    if ret==0:
        device_info = device_infolist[0]
        print("[scGetDeviceInfoList] success, ScStatus({}). The first deviceInfo, <serialNumber>: {} , <ip>: {} , <status>: {}".format(ret, str(device_info.serialNumber.decode()), str(device_info.ip.decode()), str(device_info.status)))
        if  ScConnectStatus.SC_CONNECTABLE.value != device_info.status:
            print(" The first device [status]: {} does not support connection.".format(str(device_info.status)))
            exit(1)
    else:
        print("[scGetDeviceInfoList] fail, ScStatus({})".format(ret))
        exit(1) 
else: 
    print("there are no camera found")
    exit(1)

ret = camera.scOpenDeviceBySN(device_info.serialNumber)
if  ret == 0:
    print('[scOpenDeviceBySN] success ScStatus({}).'.format(str(ret)))
else:
    print('[scOpenDeviceBySN] fail ScStatus({}).'.format(str(ret)))
    exit(1)

ret, frameRate = camera.scGetFrameRate()
if  ret == 0:
    print('[scGetFrameRate] success ScStatus({}). The device frame rate is {}.'.format(str(ret), frameRate))
else:
    print('[scGetFrameRate] fail ScStatus({}).'.format(str(ret)))
    exit(1)

ret = camera.scStartStream()
if  ret == 0:
    print('[scStartStream] success ScStatus({}).'.format(str(ret)))
else:
    print('[scStartStream] fail ScStatus({}).'.format(str(ret)))
    exit(1) 
#Get the result form AI module.
ResultCount = 100
for i in range(ResultCount):
    ret, result = camera.scAIModuleGetResult(waitTime = c_uint32(1200))
    if ret != 0:
        print('[scAIModuleGetResult] fail ScStatus({}).'.format(str(ret)))
        time.sleep(0.005)
        continue
    ret, detecList = parse_ai_result(result)
    for i, res in enumerate(detecList):
        if res.isWarning == 1:
            print(f"area {res.index} is dangerous.")

ret = camera.scStopStream()
if  ret == 0:
    print('[scStopStream] success ScStatus({}).'.format(str(ret)))
else:
    print('[scStopStream] fail ScStatus({}).'.format(str(ret)))
    exit(1)

ret = camera.scCloseDevice()
if ret == 0:
    print('[scCloseDevice] success ScStatus({}).'.format(str(ret)))
else:
    print('[scCloseDevice] fail ScStatus({}).'.format(str(ret)))
    exit(1) 

exit(0) 

File: BaseSDK/AArch64/Samples/Base/OmniGuard/SafetyCameraGetResult/SafetyCameraGetResult.cpp
#include <thread>
#include <iostream>
#include <iomanip>
#include "Scepter_api.h"
#include "Scepter_Morph_api.h"
#include <vector>

#define ResultCount 100
using namespace std;
#pragma pack(push, 1)
typedef struct
{
	uint8_t index;
	uint8_t safetyLevel;
	uint8_t isWarning;
} DetectedResult;

struct AlgHeader
{
	/*
	//Start identifer:STX A L G
	pBuf[0] = 0x02; //STX
	pBuf[1] = 0x41; //A
	pBuf[2] = 0x4C; //L
	pBuf[3] = 0x47; //G
	uint32_t startIdentifer = 0x474C4102;
	*/
	uint32_t startIdentifer;
	uint16_t sizeOfValidData;
	uint8_t  checksum;
};
#pragma pack(pop)

int ParseResult(ScAIResult result, vector<DetectedResult>& detectedResult)
{
	uint32_t usedLen = 0;

	if (result.dataLen < sizeof(AlgHeader))
	{
		cout << "dataLen is invalid. " << result.dataLen << endl;
		return 1;
	}
	// header
	const AlgHeader& algHeader = *(const AlgHeader*)(result.pResultData);
	if (0x474C4102 != algHeader.startIdentifer)
	{
		cout << "startIdentifer is invalid. " << algHeader.startIdentifer << endl;
		return 1;
	}
	usedLen += sizeof(AlgHeader);

	if (result.dataLen < (usedLen + algHeader.sizeOfValidData))
	{
		cout << "sizeOfValidData is invalid. " << algHeader.sizeOfValidData << endl;
		return 1;
	}

	// check checksum
	uint8_t checksum = 0;
	for (int i = 0; i < algHeader.sizeOfValidData; i++)
	{
		checksum += *(const uint8_t*)(result.pResultData + usedLen + i);
	}

	if (checksum != algHeader.checksum)
	{
		cout << "checkSum is invalid. " << (int)checksum << " algHeader.checksum " << (int)algHeader.checksum << endl;
		return 1;
	}

	const uint8_t* pPackageID = (const uint8_t*)(result.pResultData + usedLen);
	usedLen++;

	if (0x00 != pPackageID[0])
	{
		cout << "pPackageID is invalid " << (unsigned int)pPackageID[0] << endl;
		return 1;
	}

	uint16_t& detectedResultLen = *(uint16_t*)(result.pResultData + usedLen);
	usedLen += sizeof(uint16_t);
	uint8_t& state = *(uint8_t*)(result.pResultData + usedLen);
	usedLen += sizeof(uint8_t);

	uint8_t& errorCode = *(uint8_t*)(result.pResultData + usedLen);
	usedLen += sizeof(uint8_t);

	if (0 != state || 0 != errorCode)
	{
		cout << " Check failed state " << (unsigned int)state << " errorCode " << (unsigned int)state << " detectedResultLen " << detectedResultLen <<endl;
		return 1;
	}

	uint8_t& detectionAreaCount = *(uint8_t*)(result.pResultData + usedLen);
	usedLen += sizeof(uint8_t);

	for (size_t i = 0; i < detectionAreaCount; i++)
	{
		detectedResult.push_back(*(DetectedResult*)(result.pResultData + usedLen));
		usedLen += sizeof(DetectedResult);
	}

	if (result.dataLen > usedLen)
	{
		cout << " ignoring useless debug data " << result.dataLen << "  " << usedLen << "  " << (result.dataLen - usedLen) << endl;
	}

	return 0;
}

int main()
{
	cout << "---SafetyCameraGetResult---" << endl;

	uint32_t deviceCount;
	ScDeviceInfo* pDeviceListInfo = NULL;
	ScDeviceHandle deviceHandle = 0;
	ScStatus status = SC_OTHERS;

	status = scInitialize();
	if (status == ScStatus::SC_OK)
	{
		cout << "[scInitialize] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scInitialize] fail, ScStatus(" << status << ")." << endl;
		return 1;
	}

	status = scGetDeviceCount(&deviceCount, 3000);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scGetDeviceCount] success, ScStatus(" << status << "). The device count is " << deviceCount << endl;
	}
	else
	{
		cout << "[scGetDeviceCount] fail, ScStatus(" << status << ")." << endl;
		return 1;
	}
	if (0 == deviceCount)
	{
		cout << "[scGetDeviceCount] scans for 3000ms and then returns the device count is 0. Make sure the device is on the network before running the samples." << endl;
		return 1;
	}

	pDeviceListInfo = new ScDeviceInfo[deviceCount];
	status = scGetDeviceInfoList(deviceCount, pDeviceListInfo);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scGetDeviceInfoList] success, ScStatus(" << status << ").";
		if (SC_CONNECTABLE != pDeviceListInfo[0].status)
		{
			cout << " The first device [status]: " << pDeviceListInfo[0].status << " does not support connection." << endl;
			delete[] pDeviceListInfo;
			pDeviceListInfo = NULL;
			return 1;
		}
	}
	else
	{
		cout << "[scGetDeviceInfoList] fail, ScStatus(" << status << ")." << endl;
		delete[] pDeviceListInfo;
		pDeviceListInfo = NULL;
		return 1;
	}

	cout << " The first deviceInfo, <serialNumber>: " << pDeviceListInfo[0].serialNumber
		<< ", <ip>: " << pDeviceListInfo[0].ip << ", <status>: " << pDeviceListInfo[0].status << endl;

	status = scOpenDeviceBySN(pDeviceListInfo[0].serialNumber, &deviceHandle);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scOpenDeviceBySN] success, ScStatus(" << status << ")." << endl;
		delete[] pDeviceListInfo;
		pDeviceListInfo = NULL;
	}
	else
	{
		cout << "[scOpenDeviceBySN] fail, ScStatus(" << status << ")." << endl;
		delete[] pDeviceListInfo;
		pDeviceListInfo = NULL;
		return 1;
	}

	status = scStartStream(deviceHandle);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scStartStream] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scStartStream] fail, ScStatus(" << status << ")." << endl;
		return 1;
	}

	for (int j = 0; j < ResultCount; j++)
	{
		ScAIResult aiResult = { 0 };
		status = scAIModuleGetResult(deviceHandle, 1200, &aiResult);
		if (status == ScStatus::SC_OK)
		{
			vector<DetectedResult> detectedResult;
			ParseResult(aiResult, detectedResult);

			for (int i = 0; i < (int)detectedResult.size(); i++)
			{
				if ((uint16_t)detectedResult[i].isWarning == 1)
				{
					cout << "area " << (uint16_t)detectedResult[i].index << " is dangerous." << endl;
				}
			}
		}
		else
		{
			cout << "[scAIModuleGetResult] fail, ScStatus(" << status << ")." << endl;
			continue;
		}
	}

	status = scCloseDevice(&deviceHandle);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scCloseDevice] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scCloseDevice] fail, ScStatus(" << status << ")." << endl;
		return 1;
	}

	status = scShutdown();
	if (status == ScStatus::SC_OK)
	{
		cout << "[scShutdown] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scShutdown] fail, ScStatus(" << status << ")." << endl;
		return 1;
	}
	cout << "---end---" << endl;

	return 0;
}
